src/hufo_smoke.1
.TH "OGLC Screensaver" 1 "" "X Version 11"
.de Ds
.Sp
.nf
..
.de De
.fi
..
.SH NAME
hufo_smoke - particle smoke effect.
.SH SYNOPSIS
.B hufo_smoke
[\--root/-r]
[\--maxfps/-x \fInumber\fP]
[\--vsync/-y \fInumber\fP]
[\--dpms/-M \fInumber\fP]
[\--foreground/-f \fInumber\fP]
[\--background/-b \fInumber\fP]
.SH DESCRIPTION
From Jeremie Allard (\fIhttp://hufo.planet-d.net/codefr_us.php\fP):
"This little smoke effect is coded with a little particle system done for 'The OpenGL Challenge'." 

Ported to Linux by Tugrul Galatali.
.SH OPTIONS
.TP 8
.B \--root
Draw on the root window.
.TP 8
.B \--maxfps \fInumber\fP
Set maximum frame rate.
.TP 8
.B \--vsync \fInumber\fP
Limit redraws to specified number of vertical refreshes.  0 - 100.  Default: 1
.TP 8
.B \--dpms \fInumber\fP
Stop rendering new frames if the display is not on.  0 - 1.  Default: 1
.TP 8
.B \--foreground \fInumber\fP
Set primary (dense smoke) color using hex color codes.  Default: ffffff
.TP 8
.B \--background \fInumber\fP
Set secondary (light smoke) color using hex color codes.  Default: ffffff
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.SH SEE ALSO
.BR X (1),
.SH COPYRIGHT
Copyright \(co 2000 by Jeremie Allard and Tugrul Galatali.  

hufo_smoke is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License version 2 as
published by the Free Software Foundation.

hufo_smoke is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.SH AUTHOR
Jeremie Allard (Hufo / N.A.A.) <hufo@planet-d.net> and Tugrul Galatali <tugrul@galatali.com>



ChangeLog
******************************* Version 0.9.1  *********************************

	Added port of Pixel City, originally by Shamus Young,
http://code.google.com/p/pixelcity/
	Replaced methods deprecated by ImageMagick.

******************************* Version 0.9.0  *********************************

	Added port of drempels, originally by Ryan M. Geiss,
http://www.geisswerks.com/about_drempels.html
	Added new saver feedback by Tugrul Galatali.
	Added support for DPMS to idle the saver when the display isn't on.
	Fixed brown paper bag bug in hyperspace.
	Added two new presets to flocks inspired by Matt Richard's Wavy Gravy on
openprocessing.org
	Tweaked hufo_smoke to avoid flat flame effect.
	Fixed unknown argument segfault in lattice.
	Switched matrixview's main character array to vertex arrays, which helps
significantly on older nVidia hardware.
	hyperspace and drempels have command line arguments to disable use of
shaders in case they behave unexpectedly.
	Minor changes to fix builds on other platforms.

******************************* Version 0.8.3 *********************************

	New screen saver Lorenz contributed by Soeren Sonnenburg <sonne@debian.org>
	Fix for SF #2673397 by changing =='s to >='s in widthChange update logic,
otherwise non-sense values for interpolation would wreak havoc. Also changed for
xyzChange.

******************************* Version 0.8.2 *********************************

	Limit frame rate to vertical refresh by default if compiled with GLEW
and GLX_SGI_swap_control is available. Otherwise, maxfps set to 30.
	--nice eliminated and made default. nanosleep/usleep are much better 
behaved these days.
	spirographx modified to use vertex arrays and GNU extension sincosf if
available, dropping CPU usage significantly.
	Minor updates to correct issues noted by newer compilers.
	--disable-debug is now the default in configure.

******************************* Version 0.8.1 *********************************

	Pulled in vroot.h patch from Ari Pollak's Debian package.
	Added missing #include <assert.h> to hyperspace.cpp
	Image loading in matrixview was hanging for whatever reason,
		rewritten with the new Magick Wand interface.
	ImageMagick now required.
	timeval to double conversion now fixed with an explicit cast of tv_sec
		to double.
	Fixed automake files to use CLEANFILES
	Dropped SSE/3dnow stuff
	Fixed hyperspace.xml typo in dist_kdessconfig_DATA

******************************* Version 0.8.0 *********************************

	Rough sync against new Windows release, rssavers 0.1
		Except for skyrocket
		Added new saver Hyperspace

******************************* Version 0.7.6 *********************************

	Fix to matrixview.c where the body of a thread was accidentally left as a
subfunction of a function that exits while the thread is still alive.

	Added an autoconf check for ExportImagePixels since it was very recently
added (5.5.7).

******************************* Version 0.7.5 *********************************

	Removed multiline strings that are no longer supported by gcc 3.3.

	Added spirographx, a port of an OSX app/screensaver SpirographX by 
Holmes Futrell <holmes@neatosoftware.com>.
	http://homepage.mac.com/solidmag/spiro/

	Added matrixview, based on Alex Zolotov's Matrix GL. Hacked in external 
image loading and ability to modify the resolution of the matrix.

******************************* Version 0.7.4 *********************************

	Fixed bug in lattice when the variable passed to acos would be slightly
greater than 1, causing cascading nan's that made the program useless.

	Explicitly clearing buffers to avoid garbage in blurring modes.

	Lots of stupid little bug fixes.

******************************* Version 0.7.3 *********************************

	Added trails and blur options to flocks. Also added presets to spruce
things up in combination with random preset selection when running 
unconfigured.

	Fixed another cpu detection bug, damn off by one errors :)

	Added support for decimal integer arguments to -window-id. 

	Added .desktop files for preliminary support of KDE. Not sure where
the default directory is, but if it isn't specified by --with-kdessconfigdir,
the configure script checks for the existence of:

/usr/share/applnk/System/ScreenSavers

******************************* Version 0.7.2 *********************************

	Fixed bug in cpu detection code.

	Fixed a bug that included audio data in skyrocket when OpenAL wasn't 
found.

******************************* Version 0.7.1 *********************************

	Fixed a typo crashing skyrocket on big endian machines. Spotted by
Michel Danzer.

	Two new savers: biof and busyspheres.

	New texture smooth for colorfire.

	Fixed a bug that left screensavers running after they should have quit.

	Replaced compile time cpu feature detection with runtime detection.

******************************* Version 0.7.0 *********************************

	Extracted all binary data from the source into binary files. The
source distribution has gotten notably smaller as a result. 

	The binary files are now converted during compilation to significantly 
larger source files with escaping instead of base64 type inefficient strings.
The resulting executables have no unnecessary bloat.

	If the bzip2 library is detected on the system, it is used to compress
the binary data before conversion.

	Added --disable-bzip2 flag to configure. This is to give up space in
order to avoid a delay while decompressing, which is noticeable in skyrocket
that has about 3MB of compressed textures and audio.

	Added --disable-sound flag to configure. This chops off about 3.3MB
(before compression) of audio from skyrocket and any runtime overhead
associated with sound. Useful if you intend to keep passing volume = 0 to
skyrocket.

	Added SSE/3DNow optimized code that gives euphoria a nice 10% boost
on my P4, probably a bit more on an Athlon but I haven't had recent console 

src/hufo_smoke.cpp
/*
 * Copyright (C) 2000  Jeremie Allard (Hufo / N.A.A.)
 * Ported to Linux by Tugrul Galatali <tugrul@galatali.com>
 *
 * hufo_smoke is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as 
 * published by the Free Software Foundation.
 *
 * hufo_smoke is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

// hufo_smoke screen saver

#include <math.h>
#include <stdio.h>
#include <GL/gl.h>
#include <GL/glu.h>

#include "driver.h"
#include "rsRand.h"
#include <rsMath/rsMatrix.h>

const char *hack_name = "Hufo/N.A.A. particle system";


#include "FMotion.h"
#include "rsRand.h"

#define HVCtrX 0.0
#define HVCtrY 0.0
float FireFoc = 2.0f;

#define FireFocX FireFoc
#define FireFocY FireFoc

#define ProjEX(p) (HVCtrX+FireFocX*(p).x()/(p).y())
#define ProjEY(p) (HVCtrY+FireFocY*(p).z()/(p).y())
#define ProjEZ(ez,p) (ez)=-((p).y()*(float)(1.0/20.0)-1.0)

double nrnd (double d)		// normal distribution randomizer
{
	double r, v1, v2;

	do {
		v1 = rsRandf (2.0) - 1.0;
		v2 = rsRandf (2.0) - 1.0;
		r = v1 * v1 + v2 * v2;
	} while (r >= 1.0 || r == 0.0);

	r = sqrt (-2.0 * log (r) / r);

	return v1 * d * r;
}

#define NBPARTMAX 2048
#define PARTLIFE 2.0
#define PARTINTERV (PARTLIFE/NBPARTMAX)
#define FIRESIZE 1.0f
#define FIREDS 0.7
#define FIREALPHA 0.15f
#define FIREDA 0.4f
#define FIRECYLR 0.2f

struct Particle {
	rsVec p;		// position
	rsVec v;		// dp/dt
	rsVec s;		// size
	float a;		// alpha
	float t;		// time to death
	float ex, ey, ez;	// screen pos
	float dx, dy;		// screen size
} TblP[NBPARTMAX];

int np;
float LastPartTime;
Bool FireAnim, FireRotate, FireRecalc;
Bool FireStop;
float FireRot;
float FireAng;

rsVec FireSrc (0.0f, 25.0f, -13.0f);
rsVec FireDS1 (6.0f, 0.0f, 0.0f);
rsVec FireDS2 (0.0f, 2.0f, 0.0f);
rsVec FireDir (0.0f, 0.0f, 10.0f);
rsVec FPartA (0.0f, 0.0f, 4.0f);

rsMatrix FireM;
rsVec FireO;

void FireInit ()
{
	//NoiseInit();
	np = 0;
	LastPartTime = 0.0;
	FireAnim = True;
	FireRotate = True;
	FireRot = M_PI / 5.0;
	FireAng = 0.0;
	FireStop = False;
	FireRecalc = True;
	FMotionInit ();
/*	np=1;
	TblP[0].p=FireSrc;
	TblP[0].v.Zero();
	TblP[0].a=1.0;;
	TblP[0].s.x=2.0;
	TblP[0].s.y=2.0;
	TblP[0].s.z=2.0;
	TblP[0].t=-1;*/
}

void CalcFire (float t, float dt)
{
	int n;
	Particle *p = TblP;	//+1;

	n = 0;
	float da = pow (FIREDA, dt);
	//float ds = pow (FIREDS, dt);
	//rsVec v;

	if (FireAnim) {
		FMotionAnimate (dt);
		while (n < np) {
			if ((p->t -= dt) <= 0) {	// kill it
				*p = TblP[--np];
			} else {	// animate it
				//a=p->p; a.z-=5*t;
				//a=FPartA; //+Noise(a)*(p->p.z-FireSrc.z)*1.0; //a.x*=4.0f; a.y*=4.0f;
				p->v += FPartA * dt;
				//GetFMotion(p->p,a); p->v=a+FPartA;
				p->p += p->v * dt;
				//p->p+=FireDir*dt;
				//p->p.z+=dt;
				FMotionWarp (p->p, dt);
				//p->s*=ds;
				//p->s.x=0.67*FIRESIZE+0.4*FIRESIZE*sin(PI*(p->t)/PARTLIFE);
				//p->s.z=0.05*FIRESIZE+0.768*FIRESIZE*(1.0-p->t/PARTLIFE);
				p->a *= da;
				++n;
				++p;
			}
		}
		if (!FireStop)
			while (np < NBPARTMAX && t - LastPartTime >= PARTINTERV) {
				LastPartTime += (float)PARTINTERV;
				p = TblP + (np++);
				p->p = FireSrc + FireDS1 * nrnd (0.25) + FireDS2 * nrnd (0.25);
				p->v = FireDir;
				FMotionWarp (p->p, (t - LastPartTime));
				p->t = PARTLIFE + LastPartTime - t;
				float size = FIRESIZE * (0.5f + nrnd (0.5));
				float alpha = FIREALPHA * (float)pow (size / FIRESIZE, 0.5f);

				p->s.x() = size;
				p->s.y() = size;
				p->s.z() = size;
				p->a = alpha * pow (FIREDA, (t - LastPartTime));
				p->s *= 0.5f + nrnd (0.5);
			}
	} else
		LastPartTime += dt;

	n = 0;
	p = TblP;
	rsVec v;

	if (FireRotate) {
		FireAng += FireRot * dt;
		FireRecalc = True;
	}

	FireM.makeRot (FireAng, 0.0, 0.0, 1.0);
	FireO = FireSrc;
	FireO.transVec(FireM);
	FireO = FireSrc - FireO;
	while (n < np) {
		v = p->p;
		v.transVec(FireM);
		v += FireO;
		v = p->p;
		if (v.y() > 1.0f) {
			p->ex = ProjEX (v);
			p->ey = ProjEY (v);
			ProjEZ (p->ez, v);
			p->dx = FireFocX * p->s.x() / v.y();
			p->dy = FireFocY * p->s.z() / v.y();
		} else
			p->dx = 0;
		++n;
		++p;
	}
}

#define XSTD (4.0/3.0)
float TFire;			// fire time

bool AffGrid = false;
float fr = 1.0f, fg = 1.0f, fb = 1.0f;
float br = 1.0f, bg = 1.0f, bb = 1.0f;

void hack_reshape (xstuff_t * XStuff)
{
	float x = (float)XStuff->windowWidth / (float)XStuff->windowHeight;	// Correct the viewing ratio of the window in the X axis.

	// Window initialization
	glViewport (0, 0, XStuff->windowWidth, XStuff->windowHeight);

	glMatrixMode (GL_PROJECTION);
	glLoadIdentity ();
	if (x > XSTD)
		gluOrtho2D (-x, x, -1, 1);	// Reset to a 2D screen space.
	else
		gluOrtho2D (-XSTD, XSTD, -XSTD / x, XSTD / x);	// Reset to a 2D screen space.
	glMatrixMode (GL_MODELVIEW);
	glLoadIdentity ();
}

// Startup Stuff.
void hack_init (xstuff_t * XStuff)	// Called right after the window is created, and OpenGL is initialized.
{
	hack_reshape (XStuff);

	glCullFace (GL_FRONT);	// reject fliped faces
	glDepthFunc (GL_LESS);

	glBlendFunc (GL_SRC_ALPHA, GL_ONE);
	glEnable (GL_BLEND);

	TFire = 0.0;
	FireInit ();		// initialise fire
}

void hack_cleanup (xstuff_t * XStuff)
{
}

#define fSQRT_3_2 0.8660254038f
void AffParticle (float ex, float ey, float dx, float dy, float a)
{
	float hdx = 0.5f * dx;
	float s32_dy = fSQRT_3_2 * dy;

	glBegin (GL_TRIANGLE_FAN);
	glColor4f (fr, fg, fb, a);
	glVertex2f (ex, ey);
	glColor4f (br, bg, bb, 0.0f);
	glVertex2f (ex - dx, ey);
	glVertex2f (ex - hdx, ey + s32_dy);
	glVertex2f (ex + hdx, ey + s32_dy);
	glVertex2f (ex + dx, ey);
	glVertex2f (ex + hdx, ey - s32_dy);
	glVertex2f (ex - hdx, ey - s32_dy);
	glVertex2f (ex - dx, ey);
	glEnd ();
}

// Draw all the scene related stuff.
void hack_draw (xstuff_t * XStuff, double currentTime, float frameTime)
{
	TFire += frameTime;
	CalcFire (TFire, frameTime);	// animate the fire

	glClear (GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

	int n;

	glEnable (GL_BLEND);
	Particle *p = TblP;

	n = np;
	while (n) {
		if (p->dx)
			AffParticle (p->ex, p->ey, p->dx, p->dy, p->a);

		++p;
		--n;
	}

	glDisable (GL_BLEND);

	if (AffGrid) {
		glColor3f (1.0f, 1.0f, 1.0f);
		AffFMotion ();
	}
}

void hack_handle_opts (int argc, char **argv)
{
	int change_flag = 0;

	while (1) {
		int c;

#ifdef HAVE_GETOPT_H
		static struct option long_options[] = {
			{"help", 0, 0, 'h'},
			DRIVER_OPTIONS_LONG 
			{"foreground", 1, 0, 'f'},
			{"background", 1, 0, 'b'},
			{0, 0, 0, 0}
		};

		c = getopt_long (argc, argv, DRIVER_OPTIONS_SHORT "hf:b:", long_options, NULL);
#else
		c = getopt (argc, argv, DRIVER_OPTIONS_SHORT "hf:b:");
#endif
		if (c == -1)
			break;

		switch (c) {
		DRIVER_OPTIONS_CASES case 'h':
			printf ("%s:"
#ifndef HAVE_GETOPT_H
				" Not built with GNU getopt.h, long options *NOT* enabled."
#endif
				"\n" DRIVER_OPTIONS_HELP "\t--foreground/-f\n" "\t--background/-b\n", argv[0]);
			exit (1);
		case 'f':{
				int color = strtol_minmaxdef (optarg, 16, 0, 16777216, 0, 0xffffff, "--foreground: ");

				change_flag = 1;

				fr = ((color & 0xFF0000) >> 16) / 256.0f;
				fg = ((color & 0xFF00) >> 8) / 256.0f;
				fb = (color & 0xFF) / 256.0f;
			}
			break;
		case 'b':{
				int color = strtol_minmaxdef (optarg, 16, 0, 16777216, 0, 0xffffff, "--background: ");

				change_flag = 1;

				br = ((color & 0xFF0000) >> 16) / 256.0f;
				bg = ((color & 0xFF00) >> 8) / 256.0f;
				bb = (color & 0xFF) / 256.0f;
			}
			break;
		}
	}

	if (!change_flag) {
		do {
			fr = rsRandf (256) / 256;
			fg = rsRandf (256) / 256;
			fb = rsRandf (256) / 256;

			br = rsRandi (256) / 256;
			bg = rsRandi (256) / 256;
			bb = rsRandi (256) / 256;
		} while ((fr + fg + fb < 1) && (br + bg + bb < 1));
	}
}
